physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,583 @@
+	return d.Duration().String()
+}
+
+func (d *ReadableDuration) Duration() time.Duration {
+	if d == nil {
+		return time.Duration(0)
+	}
+	return time.Duration(*d)
+}
+
+func (d *ReadableDuration) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%s"`, d.Duration().String())), nil
+}
+
+func (d *ReadableDuration) UnmarshalJSON(raw []byte) (err error) {
+	if d == nil {
+		return fmt.Errorf("cannot unmarshal to nil pointer")
+	}
+
+	var dur time.Duration
+	str := string(raw)
+	if len(str) >= 2 && str[0] == '"' && str[len(str)-1] == '"' {
+		// quoted string
+		dur, err = time.ParseDuration(str[1 : len(str)-1])
+		if err != nil {
+			return err
+		}
+	} else {
+		// no quotes, not a string
+		v, err := strconv.ParseFloat(str, 64)
+		if err != nil {
+			return err
+		}
+		dur = time.Duration(v)
+	}
+
+	*d = ReadableDuration(dur)
+	return nil
+}
+
+func stringIDs(ids []raft.ServerID) []string {
+	out := make([]string, len(ids))
+	for i, id := range ids {
+		out[i] = string(id)
+	}
+	return out
+}
+
+func autopilotToAPIHealth(state *autopilot.State) *AutopilotState {
+	out := &AutopilotState{
+		Healthy:          state.Healthy,
+		FailureTolerance: state.FailureTolerance,
+		Leader:           string(state.Leader),
+		Voters:           stringIDs(state.Voters),
+		Servers:          make(map[string]*AutopilotServer),
+	}
+
+	for id, srv := range state.Servers {
+		out.Servers[string(id)] = autopilotToAPIServer(srv)
+	}
+
+	autopilotToAPIStateEnterprise(state, out)
+
+	return out
+}
+
+func autopilotToAPIServer(srv *autopilot.ServerState) *AutopilotServer {
+	apiSrv := &AutopilotServer{
+		ID:          string(srv.Server.ID),
+		Name:        srv.Server.Name,
+		Address:     string(srv.Server.Address),
+		NodeStatus:  string(srv.Server.NodeStatus),
+		LastContact: NewReadableDuration(srv.Stats.LastContact),
+		LastTerm:    srv.Stats.LastTerm,
+		LastIndex:   srv.Stats.LastIndex,
+		Healthy:     srv.Health.Healthy,
+		StableSince: srv.Health.StableSince,
+		Status:      string(srv.State),
+		Meta:        srv.Server.Meta,
+	}
+
+	autopilotToAPIServerEnterprise(srv, apiSrv)
+
+	return apiSrv
+}
+
+// GetAutopilotServerState retrieves raft cluster health from autopilot to
+// return over the API.
+func (b *RaftBackend) GetAutopilotServerState(ctx context.Context) (*AutopilotState, error) {
+	b.l.RLock()
+	defer b.l.RUnlock()
+
+	if b.raft == nil {
+		return nil, errors.New("raft storage is not initialized")
+	}
+
+	if b.autopilot == nil {
+		return nil, nil
+	}
+
+	return autopilotToAPIHealth(b.autopilot.GetState()), nil
+}
+
+func (b *RaftBackend) setupAutopilot(opts SetupOpts) error {
+	// Start with a default config
+	b.autopilotConfig = b.defaultAutopilotConfig()
+
+	// Check if the config was present in storage
+	switch opts.AutopilotConfig {
+	case nil:
+		// Autopilot config wasn't found in storage. Check if autopilot settings were part of config file.
+		conf, err := b.autopilotConf()
+		if err != nil {
+			return err
```

Comment:
Maybe if we have problems parsing the autopilot config we should log and fall back to the default config?  It goes against my own instincts but seems more consistent with how Vault behaves generally.

Reply:
Very good point. Thanks for this comment. This can avoid unnecessary pain if we do encounter this case. Fixed.